File: Sample.java
//1.Level order traversal
//Using DFS
//Time Complexity -> O(n)
//Space Complexity -> O(h)
class Solution {
    //Using Depth First Search
    ;
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList();
        levelOrderHelper(root,0,result);
        return result;
        
    }

    private void levelOrderHelper(TreeNode root,int level,List<List<Integer>> result){
        //Base case
        if(root == null) return;

        //recursive case
        if(result.size()==level){
            result.add(new ArrayList());
        }       
        result.get(level).add(root.val);
        level = level + 1;
        levelOrderHelper(root.left,level,result);
        levelOrderHelper(root.right,level,result);



    }
}

//Using BFS
//Time Complexity -> O(n)
//Space Complexity -> O(n/2) -> O(n) of the queue
class Solution {
    //Level Order Traversal Using Breadth First Search(BFS)
    public List<List<Integer>> levelOrder(TreeNode root) {        
        List<List<Integer>> result = new ArrayList();
        if(root == null){
            return  result;
        }
        Queue<TreeNode> q = new LinkedList(); 
        q.add(root);
        while(!q.isEmpty()){
            int size = q.size(); 
            List<Integer> li = new ArrayList();          
            for(int i=0;i<size;i++){
                 TreeNode curr = q.poll();
                 li.add(curr.val);
                 if(curr.left!=null){
                     q.add(curr.left);                     
                 }
                 if(curr.right!=null){
                     q.add(curr.right);
                 }
            }
            result.add(li);
        }
        return result;

    }
            
}

//2nd Course Schedule
//used topological sorting
//Time Complexity -> O(V+E)
//Space Complexity -> O(V + E) of the map,queue -> O(V) and indegrees -> O(V)
class Solution {
    //Using BFS
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        if(prerequisites == null || prerequisites.length == 0) return true;
        //dependent->pr[0] -> indegree and independent -> pr[1] -> outdegree
        Map<Integer,List<Integer>> map = new HashMap();
        int[] inDegrees = new int[numCourses];
        for(int[] pr : prerequisites){
            inDegrees[pr[0]] ++; //adding + 1 to indegrees of vertices on dependent nodes i.e pr[0] 
            if(!map.containsKey(pr[1])){
                map.put(pr[1],new ArrayList());
            }
            map.get(pr[1]).add(pr[0]);
        }

        Queue<Integer> q = new LinkedList();
        int count  = 0;
        for(int i = 0;i< numCourses;i++){
            if(inDegrees[i] == 0){
                q.add(i);
                count++;
            }
        }
        if(q.isEmpty()) return false; //it means there bo independent course at the starting ,hence a deadlock occured therefore return false
        if(count == numCourses) return true;
        while(!q.isEmpty()){
            int course = q.poll();
            List<Integer> children = map.get(course);
            if(children!=null){
                for(int child : children){
                    inDegrees[child]--;
                    if(inDegrees[child] == 0){
                        q.add(child);
                        count ++;
                        if(count == numCourses) return true;
                    }
                }
            }
        }

        return false;
        
    }
}

File: sample.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    
    """
    Description: Given a binary tree, return the level order traversal of its nodes' values. (ie, from left to right, level by level)
    
    Time Complexicity: O(n), n -> number of items in tree
    Space Complexicity: O(2^h) or O(n), where h -> height of the tree
    
    Approach:
    1. Use BSF -> create an empty queue
    2. add values at each level under while loop by popping the queue and traversing through both left and right nodes
    
    """
    # Using BSF
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        
        from collections import deque
        
        result = []
        if root == None: return result
        
        queue = deque()
        queue.append(root)
        
        while len(queue) != 0:
            temp = [];
            for i in range(len(queue)):
                currNode = queue.popleft()
                temp.append(currNode.val)
                if currNode.left: queue.append(currNode.left)
                if currNode.right: queue.append(currNode.right)
            result.append(temp)
            
        return result
		
class Solution:
    
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        
        """
        Description: Given the total number of courses and a list of prerequisite pairs
        
        Time Complexicity: O(n)
        Space Complexicity: O(n)
        
        Approach:
        1. Create a lead degree array and adjucancy hashmap
        2. add items to queue (to check dependency)
        """       
	
        from collections import deque
        
        if len(prerequisites) == 0: return True
        ndegrees = [0]*numCourses; count = 0
        course_dict = {i: [] for i in range(numCourses)}; 
        
        for edge in prerequisites:
            course_dict[edge[1]].append(edge[0])
            ndegrees[edge[0]] += 1
        
        queue = deque()
        for i, course in enumerate(ndegrees):
            if course == 0: queue.append(i)

        while queue:
            curr = queue.popleft(); count += 1
            for k in course_dict[curr]:
                ndegrees[k] -= 1
                if ndegrees[k] == 0: queue.append(k)

        return count == numCourses

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    
    """
    Description: Given a binary tree, return values from top to bottom looking from right side of it
    
    Time Complexicity: O(n) {using either BFS or DFS}
    Space Complexicity: O(n) or O(h) {BFS and DFS respectively}
    
    Approach:
    Using BFS:
    Use level order traversal of the binary tree and return right most node
    append the node to a resulting list
    
    Using DFS:
    
    """
    
    # Solution Using BFS
    def rightSideView(self, root: TreeNode) -> List[int]:
    
        from collections import deque
        
        result = []
        if root == None: return result
        
        queue = deque()
        queue.append(root)
        
        while len(queue) != 0:
            temp = [];
            for i in range(len(queue)):
                currNode = queue.popleft()
                temp.append(currNode.val)
                if currNode.left: queue.append(currNode.left)
                if currNode.right: queue.append(currNode.right)
            result.append(temp[-1])
            
        return result
    
    # Solution Using DFS
    def rightSideView(self, root: TreeNode) -> List[int]:
        
        self.result = []
        self.dfs(root, 0)
        
        return self.result
    
    def dfs(self, root, depth):

        # Base case
        if root == None: return
        # Logic
        if depth == len(self.result):
            self.result.append(root.val)
        else:
            self.result[depth] = root.val
        
        self.dfs(root.left, depth + 1)
        self.dfs(root.right, depth + 1)
